[vulkan] Conv2d with optional clamp

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39410 [vulkan] Remove unused variables
* #39384 [vulkan] Shaders caching
* #39282 [vulkan] jit passes for vulkan conv2 prepack and fuse with clamp
* **#39115 [vulkan] Conv2d with optional clamp**
* #39078 [vulkan] addmm support non-vulkan inputs

Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

IvanKobzarev committed:
[vulkan] Conv2d with optional clamp

[ghstack-poisoned]

IvanKobzarev committed:
drop batch norm stub as fuse_models Conv2-BatchNorm2 works on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

[ghstack-poisoned]

IvanKobzarev committed:
rebase on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -218,13 +220,17 @@ void conv2d_depthwise(
-                {params.W, params.H, params.C_4, 0}};
+                {params.W, params.H, params.C_4, 0},
+                output_min ? *output_min : std::numeric_limits<float>::min(),
+                output_max ? *output_max : std::numeric_limits<float>::max()};
```
These should be -std::numeric_limits<float>::infinity() and +std::numeric_limits<float>::infinity() respectively.

std::numeric_limits<float>::min() returns FLT_MIN which is the minimum normalized positive value of float.  Pretty confusing naming.  std::numeric_limits<float>::max() returns FLT_MAX which represents the maximum finite value of float - okay in practice but infinity is preferable in theory.

Considering that all GPUs (I am aware of) implement the IEEE 754 floating-point standard, they should understand these bit representations.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -390,13 +398,21 @@ void conv2d(
+  float outputMin =
+      output_min ? *output_min : std::numeric_limits<float>::min();
+  float outputMax =
+      output_max ? *output_max : std::numeric_limits<float>::max();
```
Same with min and max.

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -447,46 +464,71 @@ void conv2d(
     c10::optional<float*> bias,
-    const Conv2DParams params) {
-  conv2d(output, input, *(weight_prepacked.image()), bias, params);
+    const Conv2DParams params,
```
Consider passing this by const ref.  Sorry I missed that the first time.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -447,46 +464,71 @@ void conv2d(
     const VulkanTensor& bias,
-    const Conv2DParams params) {
-  conv2d(output, input, *(weight_prepacked.image()), *(bias.buffer()), params);
+    const Conv2DParams params,
```
Same here and below.

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Conv2d with optional clamp"


Conv2d with optional clamp, to handle fused conv2d_clamp from prepacking

Differential Revision: [D21962427](https://our.internmc.facebook.com/intern/diff/D21962427)

[ghstack-poisoned]
